doc/mkproject.in.man
.\" Originally generated by cmd.
.TH MKPROJECT "1" "Monday July 2014" "mkproject 2014.06.07" "User Commands"
.SH NAME
mkproject \- Project Generator
.SH SYNOPSIS
.B mkproject [OPTIONS]
.SH EXAMPLE
.B mkproject --skeleton bash --tarname test
.SH DESCRIPTION
make project is a command that makes project skeletons. mkproject automatizes the task of starting a new project with the information provided from the command line. The package created by default is a 'hello world' project of the selected skeleton that is managed with autotools. There are skeletons for bash, c, c library, python, c++, c++ library and perl.
.PP
Mandatory arguments for long options are mandatory for short options too.
.TP
\fB\-l\fR, \fB\-\-list\fR
list skeletons
.TP
\fB\-s\fR, \fB\-\-skeleton\fR=\fISKELETON\fR
set skeleton type
.TP
\fB\-t\fR, \fB\-\-tarname\fR=\fITARNAME\fR
set tar name
.TP
\fB\-p\fR, \fB\-\-pkgname\fR=\fIPKGNAME\fR
set package name
.TP
\fB\-d\fR, \fB\-\-description\fR=\fIDESCRIPTION\fR
set short description
.TP
\fB\-a\fR, \fB\-\-author\fR=\fIAUTHOR\fR
set author
.TP
\fB\-b\fR, \fB\-\-bugreport\fR=\fIBUGREPORT\fR
set bug report address
.TP
\fB\-m\fR, \fB\-\-homepage\fR=\fIHOMEPAGE\fR
set homepage url
.TP
\fB\-n\fR, \fB\-\-dependencies\fR=\fIDEPENDENCIES\fR
set dependencies
.TP
\fB\-r\fR, \fB\-\-verbose\fR
show progress messages
.TP
\fB\-v\fR, \fB\-\-version\fR
show the program version
.TP
\fB\-v\fR, \fB\-\-variable\fR=\fIVARIABLE\fR
show a program variable
.TP
\fB\-h\fR, \fB\-\-help\fR
show the program help
.TP
\fB\-m\fR, \fB\-\-man\fR
show the program manual
.TP
Skeleton and Tar Name are obligatory, the rest are recommended, dependencies is optional.
.TP
The available --variable VARIABLES are: package, name, version, description, explanation, license, homepage, author, email, social, blog, usage, options, examples, extrahelp, extranotes, month, year, day and timestamp.
.TP
Execute 'bash -c "man <(mkproject --man)"' to see the runtime manpage.
.SH NOTES
For more information, check documentation.
.SH HOMEPAGE
Main project page at \fBhttp://code.google.com/p/codemiscs\fR.
.SH AUTHOR
Programmed by \fBJuan Manuel Borges Caño\fR.
.SH EMAIL
Thank, Contact, Follow through \fBjuanmabcmail@gmail.com\fR.
.SH SOCIAL
Interact with \fBhttp://plus.google.com/+JuanManuelBorgesCaño\fR
.SH SUPPORT
Support through Paypal, Patreon, Flattr, Western Union, BitCoin, ...
.SH COPYRIGHT
Copyright \(co \fB2014 Juan Manuel Borges Caño\fR.
.PP
This software is licensed and under the terms of the GPLv3+ License.
There is NO WARRANTY, to the extent permitted by law.


NEWS
2014-07-03 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* New Commands: tonize, status.
	* Improved Commands: colorize, tw.
	* Improved Website: better looking design with more easy access to stuff that matters, and more content like NEWS & ChangeLog.
	* Better Design: colorize understands pattern and line matches, removes reverse duplication, tw adapts way faster to online engines changes. Website makes core usage of templates & is better metatagged.
	* Better Documentation: continues improving effortlessly through documentation automatic syndication, automatic tonize & status man pages.
	* Better Distribution: more automated generation of documentation and wiki.
	* Better Motivation: proved benefits sweetens development, nicer web with a lot of content exploit.

2014-03-16 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* Proved Benefits: previous features consolidate actual ones.
	* Better Motivation: proved benefits sweetens development.
	* Improved Commands: cmd (core), typewrite, colorize.
	* Improved Documentation: more rich content with media.
	* Improved Distribution: more automated generation of binary packages.
	* Better Design: typewrite can scroll, and pass through non printable characters, colorize understands more keywords.
	* Better Website: thanks to better documentation and better distribution.

2014-02-10 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* Improved Commands: cmd (core), tw, show.
	* Better Website: Continues improving effortlessly through documentation automatic syndication and smart meta distribution. Demonstrations of every command.
	* Better Design: The cmd improvements benefit instantly all commands coding, look and usage experiences. Show plugin design is more modulable and easily expandable.
	* Better Documentation: Continues improving effortlessly through documentation automatic syndication. Code parchments got everywhere where matters.
	* Better Distribution: Introducing smart meta distribution technology (currently targeted to open build service, build.opensuse.org instance, supporting fedora, ubuntu, suse, debian, centos, arch and autotools).
	* Proved Benefits: previous releases features consolidate into real value.
	* Better Motivation: copyrighted for 2014.

2013-12-30 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* New Commands, Improved Commands, Better Design, Better Documentation, Better Motivation.

2013-12-19 codemiscs <code.google.com/p/codemiscs>
	* New Commands, Better Documentation, Better Website, Better Distribution.

2013-11-28 codemiscs <code.google.com/p/codemiscs>
	* Independent Projects Merger.


cmds/mkproject/mkproject.in.sh
#! /usr/bin/env bash

#  ______________________________________________________________________________
# /\                                                                             \
# \_|         ___        __                            _   _                     |
#   |        |_ _|_ __  / _| ___  _ __ _ __ ___   __ _| |_(_) ___  _ __          |
#   |         | || '_ \| |_ / _ \| '__| '_ ` _ \ / _` | __| |/ _ \| '_ \         |
#   |         | || | | |  _| (_) | |  | | | | | | (_| | |_| | (_) | | | |        |
#   |        |___|_| |_|_|  \___/|_|  |_| |_| |_|\__,_|\__|_|\___/|_| |_|        |
#   |                                                                            |
#   |                mkproject (Make Project): Project Generator                 |
#   |             Copyright (C) 2007 - 2014 Juan Manuel Borges Caño              |
#   | The repetition of project generation made this command a need, and a must. |
#   |                      _     _                                               |
#   |                     | |   (_) ___ ___ _ __  ___  ___                       |
#   |                     | |   | |/ __/ _ \ '_ \/ __|/ _ \                      |
#   |                     | |___| | (_|  __/ | | \__ \  __/                      |
#   |                     |_____|_|\___\___|_| |_|___/\___|                      |
#   |                                                                            |
#   |    This program is free software: you can redistribute it and/or modify    |
#   |    it under the terms of the GNU General Public License as published by    |
#   |     the Free Software Foundation, either version 3 of the License, or      |
#   |                    (at your option) any later version.                     |
#   |                                                                            |
#   |      This program is distributed in the hope that it will be useful,       |
#   |       but WITHOUT ANY WARRANTY; without even the implied warranty of       |
#   |       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        |
#   |                GNU General Public License for more details.                |
#   |                                                                            |
#   |     You should have received a copy of the GNU General Public License      |
#   |   along with this program.  If not, see <http://www.gnu.org/licenses/>.    |
#   |   _________________________________________________________________________|_
#    \_/___________________________________________________________________________/

# The --list option
function mp_list
{
	shopt -s nullglob

	for mp_skeleton in "$cmd_datadir/skeletons"/*.sh
	do
		mp_skeleton="${mp_skeleton##*/}"
		mp_skeleton="${mp_skeleton%%.*}"
		printf "%s\n" "$mp_skeleton"
	done
	exit 0
}

# The --skeleton option
function mp_skeleton
{
	mps_skeleton="$1"
}

# The --tarname option
function mp_tarname
{
	mps_tarname="$1"
}

# The --pkgname option
function mp_pkgname
{
	mps_pkgname="$1"
}

# The --description option
function mp_description
{
	mps_description="$1"
}

# The --author option
function mp_author
{
	mps_author="$1"
}

# The --bugreport option
function mp_bugreport
{
	mps_bugreport="$1"
}

# The --homepage option
function mp_homepage
{
	mps_homepage="$1"
}

# The --dependencies option
function mp_dependencies
{
	mps_dependencies="$1"
}

# The --verbose option
function mp_verbose
{
	mp_verbose="on"
}

# ... and mkproject, the program itself

# The cmd init function
function mp_init
{
	mps_skeleton=""
	mps_tarname=""
	mps_pkgname=""
	mps_description=""
	mps_author=""
	mps_bugreport=""
	mps_homepage=""
	mps_dependencies=""
	mp_verbose="off"
}

# The cmd main function
function mp_main
{
	if [[ -z "$mps_skeleton" ]]
		then cmd_error "skeleton not specified"
	else
		if [[ ! -f "$cmd_datadir/skeletons/$mps_skeleton.sh" ]]
			then cmd_error "unknown skeleton"
		fi
	fi
	if [[ -z "$mps_tarname" ]]
		then cmd_error "tar name not specified"
	fi
	if [[ -z "$mps_pkgname" ]]
		then mps_pkgname="$mps_tarname"
	fi
	if [[ -z "$mps_description" ]]
		then mps_description="@description@"
	fi
	if [[ -z "$mps_author" ]]
		then mps_author="$(grep "$(whoami)" /etc/passwd | cut -d':' -f5 | cut -d',' -f1)"
		if [[ -z "$mps_author" ]]
			then mps_author="@author@"
		fi
	fi
	if [[ -z "$mps_bugreport" ]]
		then mps_bugreport="@bugreport@"
	fi
	if [[ -z "$mps_homepage" ]]
		then mps_homepage="@homepage@"
	fi

	mps_date="$(LANG=C date "+%F")"

	source "$cmd_datadir/mps.sh"
	source "$cmd_datadir/skeletons/$mps_skeleton.sh"
}

# The cmd fields
cmd_package="[@]pkg[@]"
cmd="mkproject"
cmd_name="make project"
cmd_version="[@]pkgversion[@]"
cmd_description="Project Generator"
cmd_explanation="make project is a command that makes project skeletons. mkproject automatizes the task of starting a new project with the information provided from the command line. The package created by default is a 'hello world' project of the selected skeleton that is managed with autotools. There are skeletons for bash, c, c library, python, c++, c++ library and perl."
cmd_license="[@]pkglicense[@]"
cmd_homepage="[@]pkghomepage[@]"
cmd_author="[@]pkgauthor[@]"
cmd_email="[@]pkgemail[@]"
cmd_social="[@]pkgsocial[@]"
cmd_blog="[@]pkgblog[@]"
cmd_usage="$cmd [OPTIONS]"
cmd_options=("/l/list/list skeletons/mp_list/" "/s:/skeleton:/set skeleton type/mp_skeleton/SKELETON/" "/t:/tarname:/set tar name/mp_tarname/TARNAME/" "/p:/pkgname:/set package name/mp_pkgname/PKGNAME/" "/d:/description:/set short description/mp_description/DESCRIPTION/" "/a:/author:/set author/mp_author/AUTHOR/" "/b:/bugreport:/set bug report address/mp_bugreport/BUGREPORT/" "/m:/homepage:/set homepage url/mp_homepage/HOMEPAGE/" "/n:/dependencies:/set dependencies/mp_dependencies/DEPENDENCIES/" "/r/verbose/show progress messages/mp_verbose/")
cmd_examples=("$cmd --skeleton bash --tarname test")
cmd_extrahelp="Skeleton and Tar Name are obligatory, the rest are recommended, dependencies is optional."
cmd_extranotes="For more information, check documentation."
cmd_init="mp_init"
cmd_main="mp_main"

cmd_datadir="[@]pkgdatadir[@]/$cmd"

# The cmd environment
source "[@]pkgdatadir[@]/cmd.sh"
